### ISSUE TITLE
Discuss: The value and future of CHANGELOG

### ISSUE THREAD

**Issue by mhammond** [mozilla/uniffi-rs#1458]
The discussion in #1445 has highlighted a broader question: Does our current CHANGELOG provide value to anyone? @bendk and I had a chat this topic and thought it's worth broadening the discussion. The problems we see with it for the "typical UniFFI consumer" are:

* It's not supplied with releases - anyone hoping to read it must note there's a new release, then navigate to the source repo, find the file and find the entries they care about.
* It's a mismatch of new features you might care about, breaking changes you must care about if you are impacted, and internal changes you probably do not care about unless you work on UniFFI itself.
* Even if you find the breaking changes you must care about, there's often no advice given for how to fix what is broken. These entries are often terse and consumers might be confused about whether it impacts them or not.

# Possible Solutions
## Do nothing
While it's not perfect, maybe it's good enough? Are users really struggling with this, or is it a problem we are inventing?

## Change what we record in CHANGELOG
We could have a policy that only "notable" changes are listed - eg, only new features which have an entry in the manual, and only breaking changes where we make more effort to describe what action is needed to unbreak.

The main issue with this is that it doesn't help increase the visibility of the CHANGELOG itself - but maybe just adding a link from the manual to the rendered file in the repo is OK?

## Record a better description of the changes in the manual
If we think getting detailed information in a new section of the manual itself is the best outcome, a process might look something like:

* Ensure that every release gets a human readable description of things to note in that release. For many releases this will be very small but for changes like the forthcoming 0.23 it would be significant.
* All new features of note are called out - ideally the description here will be small as they can just point at the new docs for that feature.
* All changes which are likely to break either consumers or anyone maintaining external bindings should be called out - and in all cases, we should explain what specific action impacted users should take (ie, a guide to upgrading)
* Internal changes, or "breaking" changes which only impact the uniffi-sub-crates are _not_ called out in this section - people can use the git log for this.

In terms of maintaining this section, maybe something like:

* Patch authors are encouraged to add their own entries to the manual, but we don't insist on this (it may be unclear if the change should even be added to the manual, they might not be comfortable with their English proficiency, etc) - if they prefer, they can just add a note to the repurposed CHANGELOG.
* CHANGELOG gets renamed to $SOMETHING and acts more like a todo-list for the person making the next release. So it might end up with notes added by a UniFFI maintainer (eg, "* Next release, be sure to $DO_SOMETHING") and notes about things that we _might_ want to include in the manual for the next release (ie, the things patch authors added instead of updating the manual directly)
* The person making the release must look over this file and do whatever is necessary (ie, add a nice description to the manual or decide it's not worth recording), then as part of the release clear out the pending entries.

This is somewhat orthogonal to #1453, but would be improved if that was implemented - it would mean the manual doesn't yet talk about the breaking changes for unreleased versions, which might confuse people.

## Something else?
The above seems quite complicated - other options?

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/UNIFFI-231)
┆Issue Number: UNIFFI-231

**Comment by badboy** [comment]
(side-note: heh, this is an ADR PR would have been easier to comment on. I wish GitHub would allow me more easily to comment on parts of comments...)

> It's not supplied with releases - anyone hoping to read it must note there's a new release, then navigate to the source repo, find the file and find the entries they care about.

fwiw I think dependabot is able to grep the changelog out of that file, isn't it?
If not maybe what we should do is to tag proper releases on GitHub that contains the changelog for that release. That way other tooling is definitely able to grab it.
This doesn't tackle the issue of usefulness, but would improve discoverability. We can also tackle that separately from improving the content.
 
> The main issue with this is that it doesn't help increase the visibility of the CHANGELOG itself - but maybe just adding a link from the manual to the rendered file in the repo is OK?

We can also add the CHANGELOG.md itselfs as a chapter in the rendered manual ([Glean does that](https://mozilla.github.io/glean/book/appendix/changelog/sdk.html)).
Makes cross-linking even easier.

> Patch authors are encouraged to add their own entries to the manual, but we don't insist on this (it may be unclear if the change should even be added to the manual, they might not be comfortable with their English proficiency, etc) - if they prefer, they can just add a note to the repurposed CHANGELOG.

It should be on us as maintainers to call out when a change would require a changelog.

> CHANGELOG gets renamed to $SOMETHING and acts more like a todo-list for the person making the next release. So it might end up with notes added by a UniFFI maintainer (eg, "* Next release, be sure to $DO_SOMETHING") and notes about things that we _might_ want to include in the manual for the next release (ie, the things patch authors added instead of updating the manual directly)

Why not file follow-up issues?
I also believe that landing docs with the change is a much better approach. We should strive to help people in writing those docs if they are uncomfortable doing so, either as review comments, by adding to the PR directly or in an immediate follow-up issue.

There might also be the opportunity for some tooling improvements.
I know of some projects that for example require some changelog entry in the PR comment itself. Tooling grabs it from there and assembles a final changelog on release (avoiding the issue of merge conflicts because of multiple in-flight PRs touching that). Or a [changeset approach](https://github.com/changesets/changesets) with one file per changeset, which get assembled later, if we're worried about the merge-conflict overhead.

> The person making the release must look over this file and do whatever is necessary (ie, add a nice description to the manual or decide it's not worth recording), then as part of the release clear out the pending entries.

That puts a burden on the person doing the release though. That person might not have enough context to accurately write the docs for that new feature (hopefully rare case).
When it comes to them wanting to do a release they are now suddenly blocked by finding someone who can write those docs.
That's why I think either docs immediately or file an issue _and assign it_ would be a better approach.

**Comment by skhamis** [comment]
Thanks for writing this up! This got me thinking how other libraries potentially do this to get an idea. Looking at something like `rand` I see their changelog (which is quite nice):

https://github.com/rust-random/rand/blob/937320cbfeebd4352a23086d9c6e68f067f74644/CHANGELOG.md

I like the idea of:
1. Keeping a section at the top called `UNRELEASED` up-to-date as PRs get merged (similar to what @badboy mentioned) and when it's time to cut a release we move it to the version that is being cut/tagged
2. Only really recording changes consumers care about or that would affect them (fixes, additions/changed)
3. An upgrade guide that is separate from the changelog (but linked) for times like `0.23` where we might need to explain what changed and how to fix a project that was using uniffi before `0.23`, then it also makes it easier to keep all upgrade guides in one doc for future reference?

Totally open to anything though, I think keeping the changelog is important however we ultimately decide to push forward.

**Comment by bendk** [comment]
All of this seems good to me:

> * Keeping a section at the top called `UNRELEASED` up-to-date as PRs get merged (similar to what @badboy mentioned) and when it's time to cut a release we move it to the version that is being cut/tagged
> 
> * Only really recording changes consumers care about or that would affect them (fixes, additions/changed)
> 
> * An upgrade guide that is separate from the changelog (but linked) for times like `0.23` where we might need to explain what changed and how to fix a project that was using uniffi before `0.23`, then it also makes it easier to keep all upgrade guides in one doc for future reference?

> We can also add the CHANGELOG.md itselfs as a chapter in the rendered manual

I think the last thing to figure out is how to track TODO notes for things like:
  - A contributor landed a PR, but we still need to write a changelog entry for it.  (I agree that we should try to land both together, but I don't want to make that a hard requirement).
  - Releases like 0.23 where I made changes to the release process and whoever does the release needs to double-check that it works.

What about creating a tracking issue for the next release whenever we do a release?  We could add notes like this as a checklist there.

**Comment by mhammond** [comment]
> What about creating a tracking issue for the next release whenever we do a release? We could add notes like this as a checklist there.

That sounds perfect. So I guess the directly actionable things here are:

* Trivial change to CHANGELOG to create the "unreleased" section.
* A new "upgrading" section in the manual, which I guess we fill out with the notes Ben added to the changelog.
* A change to `release-process.md` saying (a) you should find the tracking issue for the release you are about to make and ensure everything is done and (b) after cutting the release close that issue (noting what the release number actually was) and open a new one (with a title saying something like "Tracking but for release NEXT" because the exact version number will not be known at that point.

Anything else? I guess I'll keep this open as a reminder to do those things.
